plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/controller/FeatureDevControllerTest.kt

```diff
@@ -0,0 +1,105 @@
+
+import com.intellij.testFramework.RuleChain
+import io.mockk.every
+import io.mockk.just
+import io.mockk.mockkStatic
+import io.mockk.runs
+import kotlinx.coroutines.test.runTest
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.mockito.kotlin.any
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.spy
+import org.mockito.kotlin.times
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.whenever
+import software.aws.toolkits.jetbrains.services.amazonq.apps.AmazonQAppInitContext
+import software.aws.toolkits.jetbrains.services.amazonq.auth.AuthController
+import software.aws.toolkits.jetbrains.services.amazonq.auth.AuthNeededStates
+import software.aws.toolkits.jetbrains.services.amazonq.messages.MessagePublisher
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.FeatureDevTestBase
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.clients.FeatureDevClient
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.messages.IncomingFeatureDevMessage
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.session.RefinementState
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.session.Session
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.storage.ChatSessionStorage
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.util.uploadArtifactToS3
+
+class FeatureDevControllerTest : FeatureDevTestBase() {
+    @Rule
+    @JvmField
+    val ruleChain = RuleChain(projectRule, disposableRule)
+
+    private lateinit var controller: FeatureDevController
+    private lateinit var messenger: MessagePublisher
+    private lateinit var chatSessionStorage: ChatSessionStorage
+    private lateinit var appContext: AmazonQAppInitContext
+    private lateinit var authController: AuthController
+    private lateinit var spySession: Session
+    private lateinit var featureDevClient: FeatureDevClient
+
+    private val tabId = "tabId"
+
+    @Before
+    override fun setup() {
+        super.setup()
+        featureDevClient = mock()
+        messenger = mock()
+        chatSessionStorage = mock()
+        appContext = mock<AmazonQAppInitContext> {
+            on { project }.thenReturn(project)
+            on { messagesFromAppToUi }.thenReturn(messenger)
+        }
+        authController = spy(AuthController())
+        spySession = spy(Session("tabId", project))
+        mockkStatic("software.aws.toolkits.jetbrains.services.amazonqFeatureDev.util.UploadArtifactKt")
+        every { uploadArtifactToS3(any(), any(), any(), any(), any()) } just runs
+
+        controller = FeatureDevController(appContext, chatSessionStorage, authController)
+    }
+
+    @Test
+    fun `test new tab opened`() {
+        val message: IncomingFeatureDevMessage.NewTabCreated = IncomingFeatureDevMessage.NewTabCreated("new-tab-created", tabId)
```

Comment:
type declaration is not needed

```suggestion
        val message = IncomingFeatureDevMessage.NewTabCreated("new-tab-created", tabId)
```